src/huggingface_hub/inference/_generated/_async_client.py

```diff
@@ -327,9 +331,35 @@ async def post(
                         continue
                     raise error
                 except Exception:
-                    await client.close()
+                    await session.close()
                     raise
 
+    async def __aenter__(self):
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.close()
+
+    def __del__(self):
+        if len(self._sessions) > 0:
+            warnings.warn(
+                "Deleting 'AsyncInferenceClient' client but some sessions are still open. "
+                "This can happen if you've stopped streaming data from the server before the stream was complete. "
+                "To close the client properly, you must call `await client.close()` "
+                "or use an async context (e.g. `async with AsyncInferenceClient(): ...`."
+            )
```

Comment:
Out of curiosity, shouldn't this just close the session as well?

Reply:
Well well well...

I got a hard time getting to make it work. The thing is that you don't know when `__del__` is called by Python's garbage collector. Since `self.close` is an async function, you need an active event loop to make the close method run -which is not guaranteed. I first tried some hacks but gave up given the complexity compared to having a proper `close` method or using an async context manager.

[This answer on stackoverflow](https://stackoverflow.com/a/54773296) also convinced me of doing things like this.